### PR TITLE
CCD-3669: Replacing SQL based search with Elastic Search

### PR DESCRIPTION
**Pull request by kiran-yenigala-hmcts**
**Before creating a pull request make sure that:**

- [ ] commit messages are meaningful and follow good commit message guidelines
- [ ] README and other documentation has been updated / added (if needed)
- [ ] tests have been updated / new tests has been added (if needed)

Please remove this line and everything above and fill the following sections:

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ ] No
```

### CODE REVIEW

File: modules/processor/src/main/java/uk/gov/hmcts/reform/migration/CaseMigrationProcessor.java [hunk 1]
```diff
@@ -36,30 +48,53 @@ public void processSingleCase(String userToken, String caseId) {
...
+        updateCase(userToken, caseDetails.getCaseTypeId(), caseDetails);
     }
 
-    public void processAllCases(String userToken, String userId) {
-        coreCaseDataService.fetchAll(userToken, userId).stream()
-            .filter(dataMigrationService.accepts())
-            .forEach(caseDetails -> updateCase(userToken, caseDetails.getId(), caseDetails.getData()));
+    public void processAllCases(String userToken, String userId, String caseType) {
+        coreCaseDataService.fetchAll(userToken, userId, caseType).stream()
+            .forEach(caseDetails -> updateCase(userToken, caseType, caseDetails));
```
Comment (danlysiak) [inline modules/processor/src/main/java/uk/gov/hmcts/reform/migration/CaseMigrationProcessor.java:56]: This is using the old SQL approach, but I don't think is being used by anything now anyway (as it shouldn't be, since the new `migrateCases()` method below is the one which is invoked by the Runner and uses ES)? So we should be able to delete this.

File: modules/processor/src/main/java/uk/gov/hmcts/reform/migration/CaseMigrationProcessor.java [hunk 2]
```diff
@@ -36,30 +48,53 @@ public void processSingleCase(String userToken, String caseId) {
...
+                    caseDetails
+                );
+                log.info("Case {} successfully updated", id);
+                migratedCases.add(id);
+            } catch (Exception e) {
+                log.error("Case {} update failed due to: {}", id, e.getMessage());
+                failedCases.add(id);
+            }
+        } else {
+            log.info("Case {} already migrated", caseDetails.getId());
```
Comment (danlysiak) [inline modules/processor/src/main/java/uk/gov/hmcts/reform/migration/CaseMigrationProcessor.java:97]: Minor - because it doesn't necessarily mean that the case is already migrated, it could also be that the service implemented some rules that make it not applicable for other reasons.
```suggestion
            log.info("Case {} does not meet criteria for migration", caseDetails.getId());
```

File: modules/processor/src/main/java/uk/gov/hmcts/reform/migration/ccd/CoreCaseDataService.java [hunk 3]
```diff
@@ -38,34 +33,35 @@ public CaseDetails fetchOne(String authorisation, String caseId) {
...
     }
 
-    private List<CaseDetails> fetchPage(String authorisation, String userId, int pageNumber) {
+    private List<CaseDetails> fetchPage(String authorisation, String userId,  String caseType, int pageNumber) {
         Map<String, String> searchCriteria = new HashMap<>();
         searchCriteria.put("page", String.valueOf(pageNumber));
-        return coreCaseDataApi.searchForCaseworker(authorisation, authTokenGenerator.generate(), userId, jurisdiction,
+        return coreCaseDataApi.searchForCaseworker(authorisation, authTokenGenerator.generate(), userId, null,
             caseType, searchCriteria);
     }
```
Comment (danlysiak) [inline modules/processor/src/main/java/uk/gov/hmcts/reform/migration/ccd/CoreCaseDataService.java:54]: These three methods are using SQL - think they can be deleted?

File: modules/processor/src/main/java/uk/gov/hmcts/reform/migration/ccd/CoreCaseDataService.java [hunk 4]
```diff
@@ -38,34 +33,35 @@ public CaseDetails fetchOne(String authorisation, String caseId) {
...
+    public CaseDetails update(String authorisation, String eventId, String eventSummary, String eventDescription, String caseType, CaseDetails caseDetails) {
+        String caseId = String.valueOf(caseDetails.getId());
         UserDetails userDetails = idamClient.getUserDetails(AuthUtil.getBearerToken(authorisation));
 
         StartEventResponse startEventResponse = coreCaseDataApi.startEventForCaseWorker(
             AuthUtil.getBearerToken(authorisation),
             authTokenGenerator.generate(),
             userDetails.getId(),
-            jurisdiction,
+            null,
```
Comment (danlysiak) [inline modules/processor/src/main/java/uk/gov/hmcts/reform/migration/ccd/CoreCaseDataService.java:64]: I think jurisdiction is required for this API - but we should have it available in `caseDetails` so can get it from there. (Could also get case type from the `caseDetails`?)

File: modules/processor/src/main/java/uk/gov/hmcts/reform/migration/ccd/CoreCaseDataService.java [hunk 5]
```diff
@@ -78,14 +74,14 @@ public CaseDetails update(String authorisation, String caseId, String eventId, S
...
-            ).data(data)
+            ).data(caseDetails.getData())
             .build();
 
         return coreCaseDataApi.submitEventForCaseWorker(
             AuthUtil.getBearerToken(authorisation),
             authTokenGenerator.generate(),
             userDetails.getId(),
-            jurisdiction,
+            null,
```
Comment (danlysiak) [inline modules/processor/src/main/java/uk/gov/hmcts/reform/migration/ccd/CoreCaseDataService.java:84]: Same as previous comment.

File: modules/processor/src/main/java/uk/gov/hmcts/reform/migration/ccd/CoreCaseDataService.java (hunk 4 above)
Comment (kiran-yenigala-hmcts) [reply modules/processor/src/main/java/uk/gov/hmcts/reform/migration/ccd/CoreCaseDataService.java:64]: done

File: modules/processor/src/main/java/uk/gov/hmcts/reform/migration/ccd/CoreCaseDataService.java (hunk 5 above)
Comment (kiran-yenigala-hmcts) [reply modules/processor/src/main/java/uk/gov/hmcts/reform/migration/ccd/CoreCaseDataService.java:84]: done

File: modules/processor/src/main/java/uk/gov/hmcts/reform/migration/CaseMigrationProcessor.java (hunk 1 above)
Comment (kiran-yenigala-hmcts) [reply modules/processor/src/main/java/uk/gov/hmcts/reform/migration/CaseMigrationProcessor.java:56]: removed this

File: modules/processor/src/main/java/uk/gov/hmcts/reform/migration/CaseMigrationProcessor.java (hunk 2 above)
Comment (kiran-yenigala-hmcts) [reply modules/processor/src/main/java/uk/gov/hmcts/reform/migration/CaseMigrationProcessor.java]: done

File: build.gradle [hunk 6]
```diff
@@ -96,6 +99,12 @@ task integration(type: Test) {
   failFast = true
 }
 
+task fortifyScan(type: JavaExec)  {
+  main = "uk.gov.hmcts.fortifyclient.FortifyClientMainApp"
+  classpath += sourceSets.test.runtimeClasspath
+  jvmArgs = ['--add-opens=java.base/java.lang.reflect=ALL-UNNAMED']
+}
```
Comment (rajaram-nagarajan) [inline build.gradle:106]: We can add the below flag to prevent the build from failing if the Fortify scan detects issues as we monitor the fortify app to fix/suppress the issues.
ignoreExitValue = true

File: config/owasp/suppressions.xml [hunk 7]
```diff
@@ -19,41 +19,9 @@
   <!--Please add all the temporary suppression under the below section-->
```
Comment (rajaram-nagarajan) [inline config/owasp/suppressions.xml:19]: @kiran-yenigala-hmcts we can have another section for false positives like the temporary section. Please refer data-store suppression file.

File: config/owasp/suppressions.xml [hunk 8]
```diff
@@ -19,41 +19,9 @@
...
   <suppress>
     <notes>Temporary Suppression</notes>
```
Comment (rajaram-nagarajan) [inline config/owasp/suppressions.xml:21]: @kiran-yenigala-hmcts We can add the jira link in the notes for the respective CVEs. This will help us to map the CVE progress.

File: build.gradle (hunk 6 above)
Comment (kiran-yenigala-hmcts) [reply build.gradle:107]: done

File: src/main/java/uk/gov/hmcts/reform/migration/CaseMigrationProcessor.java [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+
+    public void migrateCases(String caseType) {
+        validateCaseType(caseType);
+        log.info("Data migration of all cases started for case type: {}", caseType);
+        String userToken =  idamRepository.generateUserToken();
+        List<CaseDetails> listOfCaseDetails = elasticSearchRepository.findCaseByCaseType(userToken, caseType);
+        listOfCaseDetails.stream()
+            .limit(caseProcessLimit)
+            .forEach(caseDetails -> updateCase(userToken, caseType, caseDetails));
+        log.info(LOG_STRING);
```
Comment (rajaram-nagarajan) [inline src/main/java/uk/gov/hmcts/reform/migration/CaseMigrationProcessor.java:57]: @kiran-yenigala-hmcts We can convert the continuous logger statements into one single log

File: src/main/java/uk/gov/hmcts/reform/migration/CaseMigrationProcessor.java [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+
+    @Getter
+    private List<Long> failedCases = new ArrayList<>();
+
+    @Value("${case-migration.processing.limit}")
+    private int caseProcessLimit;
+
+    public void migrateCases(String caseType) {
+        validateCaseType(caseType);
+        log.info("Data migration of all cases started for case type: {}", caseType);
```
Comment (rajaram-nagarajan) [inline src/main/java/uk/gov/hmcts/reform/migration/CaseMigrationProcessor.java:51]: @kiran-yenigala-hmcts We can't exactly say all cases will be considered for migration as we have a processing limit now. So, the logger statement can be modified to highlight that.

File: src/main/java/uk/gov/hmcts/reform/migration/CaseMigrationProcessor.java [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+        } else {
+            log.info("Migrated cases: {} ", getMigratedCases());
+        }
+
+        if (getFailedCases().isEmpty()) {
+            log.info("Failed cases: NONE ");
+        } else {
+            log.info("Failed cases: {} ", getFailedCases());
+        }
+        log.info("Data migration of all cases completed");
```
Comment (rajaram-nagarajan) [inline src/main/java/uk/gov/hmcts/reform/migration/CaseMigrationProcessor.java:74]: @kiran-yenigala-hmcts We can avoid 'all' word here and can just say Data migration completed

File: src/main/resources/application.properties [hunk 12]
```diff
@@ -19,7 +19,9 @@ idam.s2s-auth.totp_secret=${DATA_STORE_IDAM_KEY:AAAAAAAAAAAAAAAB}
 idam.s2s-auth.microservice=ccd_data
 idam.s2s-auth.url=${IDAM_S2S_URL:http://localhost:4502}
 
-migration.idam.username= ${IDAM_USE}
+migration.idam.username= ${IDAM_USER}
 migration.idam.password= ${IDAM_PASS}
 migration.caseType= ${MIGRATION_CASETYPE}
-migration.caseId= ${MIGRATION_CASETID:}
```
Comment (rajaram-nagarajan) [inline src/main/resources/application.properties:25]: @kiran-yenigala-hmcts You can remove the references of migration.caseId from Readme file

File: src/test/java/uk/gov/hmcts/reform/migration/CaseMigrationProcessorTest.java [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+
+    @Mock
+    private IdamRepository idamRepository;
+
+    @Before
+    public void setUp() {
+        ReflectionTestUtils.setField(caseMigrationProcessor, "caseProcessLimit", 1);
+    }
+
+    @Test
```
Comment (rajaram-nagarajan) [inline src/test/java/uk/gov/hmcts/reform/migration/CaseMigrationProcessorTest.java:57]: @kiran-yenigala-hmcts We can add a test case to cover invalid case type scenario

File: src/main/java/uk/gov/hmcts/reform/migration/repository/ElasticSearchRepository.java [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+        String authToken = authTokenGenerator.generate();
+
+        SearchResult searchResult = coreCaseDataApi.searchCases(userToken,
+                                                                authToken,
+                                                                caseType, elasticSearchQuery.getQuery()
+        );
+
+        List<CaseDetails> caseDetails = new ArrayList<>();
+
+        if (searchResult.getTotal() > 0) {
```
Comment (rajaram-nagarajan) [inline src/main/java/uk/gov/hmcts/reform/migration/repository/ElasticSearchRepository.java:51]: @kiran-yenigala-hmcts We can add null check for searchResult and subsequentSearchResult  as it is getting the values from external (data-store) API and we never know when things change.

File: src/test/java/uk/gov/hmcts/reform/migration/repository/ElasticSearchRepositoryTest.java [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+    private AuthTokenGenerator authTokenGenerator;
+
+    @Before
+    public void setUp() {
+        elasticSearchRepository = new ElasticSearchRepository(coreCaseDataApi, authTokenGenerator, QUERY_SIZE);
+        when(authTokenGenerator.generate()).thenReturn(AUTH_TOKEN);
+    }
+
+    @Test
+    public void shouldReturnSearchResultsForCaseTypeElasticSearch() {
```
Comment (rajaram-nagarajan) [inline src/test/java/uk/gov/hmcts/reform/migration/repository/ElasticSearchRepositoryTest.java:81]: @kiran-yenigala-hmcts we can add null (for both searchResult and subsequentSearchResult) and empty (searchResult) check test cases too

**Review by rajaram-nagarajan — CHANGES_REQUESTED** [review]
@kiran-yenigala-hmcts Please address the review comments

File: config/owasp/suppressions.xml (hunk 7 above)
Comment (kiran-yenigala-hmcts) [reply config/owasp/suppressions.xml:19]: done

File: src/main/java/uk/gov/hmcts/reform/migration/CaseMigrationProcessor.java (hunk 10 above)
Comment (kiran-yenigala-hmcts) [reply src/main/java/uk/gov/hmcts/reform/migration/CaseMigrationProcessor.java:51]: removed **all** from this statement

File: src/main/resources/application.properties (hunk 12 above)
Comment (kiran-yenigala-hmcts) [reply src/main/resources/application.properties:25]: removed

File: src/test/java/uk/gov/hmcts/reform/migration/repository/ElasticSearchRepositoryTest.java (hunk 15 above)
Comment (kiran-yenigala-hmcts) [reply src/test/java/uk/gov/hmcts/reform/migration/repository/ElasticSearchRepositoryTest.java:81]: addressed

File: src/test/java/uk/gov/hmcts/reform/migration/service/DataMigrationServiceImplTest.java [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+
+@RunWith(MockitoJUnitRunner.class)
+public class DataMigrationServiceImplTest {
+
+    private DataMigrationServiceImpl service = new DataMigrationServiceImpl();
+
+    @Test
+    public void shouldReturnNullWhenAccepts() {
```
[the remark this answers is not in the document]
Comment (kiran-yenigala-hmcts) [reply src/test/java/uk/gov/hmcts/reform/migration/service/DataMigrationServiceImplTest.java:21]: renamed and addressed test case accordingly

File: src/main/java/uk/gov/hmcts/reform/migration/CaseMigrationProcessor.java (hunk 9 above)
Comment (rajaram-nagarajan) [reply src/main/java/uk/gov/hmcts/reform/migration/CaseMigrationProcessor.java:57]: @kiran-yenigala-hmcts I am suggesting something like below,
        log.info("""
                     {}
                     Data migration completed
                     {}
                     Total number of processed cases: {}
                     Total number of migrations performed: {}
                     {}""",
                 LOG_STRING, LOG_STRING,
                 getMigratedCases().size() + getFailedCases().size(),
                 getMigratedCases().size(), LOG_STRING);

File: src/test/java/uk/gov/hmcts/reform/migration/service/DataMigrationServiceImplTest.java [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+public class DataMigrationServiceImplTest {
+
+    private DataMigrationServiceImpl service = new DataMigrationServiceImpl();
+
+    @Test
+    public void shouldReturnTruForCaseDetailsPassed() {
+        CaseDetails caseDetails = CaseDetails.builder()
+            .id(1234L)
+            .build();
+        assertTrue(service.accepts().test(caseDetails));
```
Comment (rajaram-nagarajan) [inline src/test/java/uk/gov/hmcts/reform/migration/service/DataMigrationServiceImplTest.java:26]: @kiran-yenigala-hmcts We can add false test case here when caseDetails is null

File: src/main/java/uk/gov/hmcts/reform/migration/CaseMigrationProcessor.java (hunk 9 above)
Comment (kiran-yenigala-hmcts) [reply src/main/java/uk/gov/hmcts/reform/migration/CaseMigrationProcessor.java:57]: addressed

File: src/main/java/uk/gov/hmcts/reform/migration/CaseMigrationProcessor.java [hunk 18]
```diff
@@ -0,0 +1,107 @@
...
+    }
+
+    private void validateCaseType(String caseType) {
```
[the remark this answers is not in the document]
Comment (kiran-yenigala-hmcts) [reply src/main/java/uk/gov/hmcts/reform/migration/CaseMigrationProcessor.java:80]: Discussed and no need any change

**Review by rajaram-nagarajan — DISMISSED** [review]
LGTM

**Review by rajaram-nagarajan — APPROVED** [review]

**Review by rajaram-nagarajan — APPROVED** [review]

**Review by JamesCollettCGI — APPROVED** [review]
